### PR TITLE
moving run_migrations task to a dedicated_process_num 

### PR DESCRIPTION
**Pull request by prem-fissionhq**
## Summary
Plan to move run_migrations task to a dedicated_process_num (i.e on the first process_num) and not to run any other processes when run_migrations is set to True.

To handle this behavior, we added a new variable dedicated_migration_process which will be passed from the app_processes.yml for the respective pillow processes.

Ticket:
https://dimagi-dev.atlassian.net/browse/SAAS-11278

## Safety Assurance

- [ ] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [ ] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below
       > The existing users/process will not have any impact with this change. 

### Safety story
Tested the PR by following below procedure

1. Deployed HQ code of this branch in the staging env by adding branchName in the scripts/staging.yaml. 
2. Deployed Commcare Code in staging env by doing following changes. 
Added `dedicated_migration_process = True` to the case-pillow in the app-processes.yml
   ```
   case-pillow:
      num_processes: 2
      dedicated_migration_process: True
   ```
 2. Verified whether the supervisor configuration has been applied with the new change to case-pillow.
 3. Monitoring/observations 
     - On process_num=0,  for every 10secs the process is checking for any new migrations to run.
     - On process_num=1, we are receiving requests whenever a case is submitted. No requests were recieved for process_num=0.
     - While testing we have submitted both forms and cases from the staging account. When we submitted forms we saw requests in xform-pillow logs and when submitted cases we saw requests in case-pillow-1.log.  
     - In order to trigger only migrations we have created/edited reports in the Report Builder, it worked as expected. 

	This PR is safe because it will not affect any existing workflow. If `dedicated_migration_process: True` is set to the pillows in the app-processes.yml then the processes will spill the migrations and other transactions on to different process_nums else it will work as it is. The change will also affect only that particular process.

### Rollback instructions

update `dedicated_miigration_process: False` in app-processes.yml and then run update-supervisor-configs. 
If the option is removed or set to False the change will be disabled.

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by prem-fissionhq** [comment]
Commcare-cloud PR - changes that are related to this.
PR: https://github.com/dimagi/commcare-cloud/pull/4288

File: corehq/apps/userreports/pillow.py [hunk 1]
```diff
@@ -192,7 +192,6 @@ def bootstrap(self, configs=None):
             self.table_adapters_by_domain[config.domain].append(
                 self._get_indicator_adapter(config)
             )
-
```
Comment (dannyroberts) [inline corehq/apps/userreports/pillow.py:195]: nit: unrelated change

File: corehq/messaging/pillow.py [hunk 2]
```diff
@@ -73,9 +73,8 @@ def process_changes_chunk(self, changes_chunk):
         return [], errors
 
 
-def get_case_messaging_sync_pillow(pillow_id='case_messaging_sync_pillow', topics=None,
-                         num_processes=1, process_num=0,
-                         processor_chunk_size=DEFAULT_PROCESSOR_CHUNK_SIZE, **kwargs):
+def get_case_messaging_sync_pillow(pillow_id='case_messaging_sync_pillow', topics=None, process_num=0,
+                         num_processes=1, processor_chunk_size=DEFAULT_PROCESSOR_CHUNK_SIZE, **kwargs):
```
Comment (dannyroberts) [inline corehq/messaging/pillow.py:77]: nit: unrelated change?

File: corehq/pillows/case.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def transform_case_for_elasticsearch(doc_dict):
 
 
 def get_case_to_elasticsearch_pillow(pillow_id='CaseToElasticsearchPillow', num_processes=1,
-                                     process_num=0, **kwargs):
+                                    process_num=0, **kwargs):
```
Comment (dannyroberts) [inline corehq/pillows/case.py:53]: nit: unrelated/unintentional change?

**Review by dannyroberts — COMMENTED** [review]
Hey Prem this is looking really good.

The only comments I had are about changes that look like they slipped into the diff and were maybe part of a change from an earlier draft of the PR, but which now aren't adding anything. I know it's a pain, but could you look at the lines I marked and if you agree clean those up from the PR diff?

**Comment by dannyroberts** [comment]
@prem-fissionhq can you work with @shyamkumarlchauhan to put this on staging and test each of the affected queues with it both on and off? @AmitPhulera may be able to work with you as well

**Comment by prem-fissionhq** [comment]
> Hey Prem this is looking really good.
> 
> The only comments I had are about changes that look like they slipped into the diff and were maybe part of a change from an earlier draft of the PR, but which now aren't adding anything. I know it's a pain, but could you look at the lines I marked and if you agree clean those up from the PR diff?

Sure, Danny. Will work on it. thanks

**Comment by prem-fissionhq** [comment]
@snopoke as discussed over skype, this is PR that is giving the below error. 

```
ERROR: test suite for <class 'corehq.apps.userreports.tests.test_data_source_citus.DataSourceConfigurationCitusDBTest'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python3.6/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/vendor/lib/python3.6/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/vendor/lib/python3.6/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/vendor/lib/python3.6/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_data_source_citus.py", line 31, in setUpClass
    cls.adapter = get_indicator_adapter(cls.data_source)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/util.py", line 172, in get_indicator_adapter
    adapter = adapter_cls(config)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/sql/adapter.py", line 41, in __init__
    self.session_helper = connection_manager.get_session_helper(self.engine_id)
  File "/mnt/commcare-hq-ro/corehq/sql_db/connections.py", line 134, in get_session_helper
    return self._get_or_create_helper(self.get_connection_string(engine_id))
  File "/mnt/commcare-hq-ro/corehq/sql_db/connections.py", line 188, in get_connection_string
    return self._connection_string_from_django(db_alias)
  File "/vendor/lib/python3.6/site-packages/memoized.py", line 20, in _memoized
    cache[key] = value = fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/sql_db/connections.py", line 215, in _connection_string_from_django
    db_settings = settings.DATABASES[django_alias].copy()
KeyError: None
----------------------------------------------------------------------
Ran 2425 tests in 719.303s
```

**Comment by snopoke** [comment]
I'm pretty sure this is because the PR is from a fork and so the ICDS code doesn't get checked out. But I think this should fix it: https://github.com/dimagi/commcare-hq/pull/28874

**Comment by prem-fissionhq** [comment]
> I'm pretty sure this is because the PR is from a fork and so the ICDS code doesn't get checked out. But I think this should fix it: #28874

Thanks, Simon, I will update the file and let you know if ran into any issues.

**Comment by prem-fissionhq** [comment]
@snopoke the above change is already present in the docker/localsettings.py. 
@dannyroberts as you suggested I merged the latest master code to this branch. 

Thank you.
cc: @AmitPhulera

**Comment by dannyroberts** [comment]
Looks to me like Simon just merged #28874, the PR he said would fix the build, and thus that your last merge a few days ago didn't include his fix. I think you just need to merge dimagi/master into your branch again now

**Review by dannyroberts — APPROVED** [review]

**Review by shyamkumarlchauhan — APPROVED** [review]

File: corehq/ex-submodules/pillowtop/pillow/interface.py [hunk 4]
```diff
@@ -106,7 +107,13 @@ def run(self):
         pillow_logging.info("Starting pillow %s" % self.__class__)
         with configure_scope() as scope:
             scope.set_tag("pillow_name", self.get_name())
-        self.process_changes(since=self.get_last_checkpoint_sequence(), forever=True)
+        if self.is_dedicated_migration_process:
+            for processor in self.processors:
+                print(processor)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py:112]: `print` should be removed

File: corehq/apps/change_feed/consumer/feed.py [hunk 5]
```diff
@@ -163,10 +165,19 @@ def _filter_offsets(self, offsets) -> Optional[Dict[TopicPartition, int]]:
...
+                for num in range(self.num_processes)
+            ][self.process_num]
+        else:
+            if self.process_num == 0:
+                return None
+            else:
+                return [
+                    topic_partitions[num::self.num_processes]
+                    for num in range(self.num_processes - 1)
+                ][self.process_num - 1]
```
Comment (snopoke) [inline corehq/apps/change_feed/consumer/feed.py:180]: This won't return the correct breakdown since we're still breaking up the `topic_partitions` list into batches of `self.num_processes`. It should be:

```
                num_processes = self.num_processes - 1
                return [
                    topic_partitions[num::num_processes]
                    for num in range(num_processes)
                ][num_processes]
```

This code is ripe some unit tests.

Comment (prem-fissionhq) [reply corehq/apps/change_feed/consumer/feed.py:180]: thank you @snopoke, I made the changes and deploy them in staging now.
